### PR TITLE
Fix inference optimization cleanup and download temp-file races

### PR DESCRIPTION
**Pull request by Borda**
This pull request introduces several improvements and bug fixes across model optimization, file handling, and loss computation. The most significant changes enhance robustness during model optimization, ensure atomic file downloads to prevent corruption in concurrent environments, and improve tensor handling for gradient tracking and resizing.

**Model Optimization and Robustness:**
- Improved error handling in `optimize_for_inference` to ensure the model is always left in a consistent, unoptimized state if optimization fails, and refactored context management for device handling. [[1]](diffhunk://#diff-88cb2cb1c972ea4e1b9e1582f78beb4ad32bf7c10709dd837c57a2da1ce98c86R541-R551) [[2]](diffhunk://#diff-88cb2cb1c972ea4e1b9e1582f78beb4ad32bf7c10709dd837c57a2da1ce98c86R576-R580)

**File Download and Atomicity:**
- Switched temporary file creation to use `tempfile.mkstemp` in the destination directory for downloads, ensuring unique temp files and avoiding collisions in concurrent scenarios. [[1]](diffhunk://#diff-e7e35aea42eddb77a9522a1609c1dd564a2146c3197d1c7ebd6d1d1a431a8d69L11-R12) [[2]](diffhunk://#diff-e7e35aea42eddb77a9522a1609c1dd564a2146c3197d1c7ebd6d1d1a431a8d69L89-R98)
- Replaced `shutil.move` with `os.replace` for atomic file replacement after download, reducing the risk of file corruption.

**Model and Loss Computation Fixes:**
- Changed tensor access in `_resize_linear` to use `.detach()` for both weights and biases, ensuring no unwanted gradient tracking during resizing.
- Updated loss computation in `_probe_step` to always create a base loss tensor with `requires_grad=True`, ensuring proper gradient flow even when `weighted_losses` is empty.

### CODE REVIEW

File: src/rfdetr/utilities/files.py [hunk 1]
```diff
@@ -117,5 +124,10 @@ def _download_file(
...
 
-    # shutil.move handles cross-device moves (e.g. tmpfs → ext4 on Colab).
-    shutil.move(temp_filename, filename)
+    # Atomic replace in target directory.
+    try:
+        os.replace(temp_filename, filename)
+    except Exception:
+        if os.path.exists(temp_filename):
+            os.remove(temp_filename)
+        raise
```
Comment (Borda) [inline src/rfdetr/utilities/files.py:132]:
```suggestion
    os.replace(temp_filename, filename)
```
